### ISSUE TITLE
[request] add subfolder

### ISSUE THREAD

**Issue by nhantrn** [choobin/cmis#5]
Hi, I have a couple requests to make.
1. Add an option to enable subfolders within an item.
2. Add option to move the Firefox default "Save Image/Link As..." to a subitem

**Comment by choobin** [comment]
G'day,

Thanks for the feedback.

> Add an option to enable subfolders within an item.

I do not quite follow. Do you mean nested sub-folders? Like this?

![example](https://f.cloud.github.com/assets/211907/49474/60da9980-5963-11e2-96cb-093c56bcfb34.png)

If so. This is already possible. Although completely undocumented. See [issue/3](https://github.com/choobin/cmis/issues/3).

Select the sub-menu root item and use the up/down buttons to move the whole sub-menu around.

> Add option to move the Firefox default "Save Image/Link As..." to a subitem

Great idea. I will look into adding this to the next update. Perhaps a "Save Image As..." button under the "Separator" button in the Menu Item Settings dialog.

**Comment by nhantrn** [comment]
Thanks for looking into my 2nd request.

As for the subfolder thing, what I mean is that if there are sub folders within a path and it should be expanded in the menuitem. Here's a visual of the concept.

![2013-01-08_222658](https://f.cloud.github.com/assets/3202654/52717/ddb6d62c-5a25-11e2-8e90-f368a8d57074.jpg)

**Comment by choobin** [comment]
>  what I mean is that if there are sub folders within a path and it should be expanded in the menuitem.

OK. I see what you mean. I do not know about that one. Can't you create a new item at the top of a sub-menu that achieves the same functionality?

Regarding the 'Save Image As' button. Commit [569f048072](https://github.com/choobin/cmis/commit/569f048072c55d80e1142a68902176cf7fca7226) if the first step towards this goal. I need to do more testing before I send it to AMO. If you want to test it out yourself I can provide an XPI and/or document how to package and install the latest snapshot from github. See [How to test the latest update from github](https://github.com/choobin/cmis/wiki/How-to-test-the-latest-update-from-github.).

**Comment by nhantrn** [comment]
Yes you can manually add sub-folder items to a submenu, but that's not an ideal solution. Especially if a path has multiple sub-folders.

As for the latest commit, it works on my end so far without issues. But I was expecting when the default Save Image As... is relocated to a submenu, it should also be removed from the main context menu.
That's the idea, else there's no point in reduplicating that menuitem everywhere ;)

**Comment by choobin** [comment]
> Yes you can manually add sub-folder items to a submenu, but that's not an ideal solution. Especially if a path has multiple sub-folders.

I see. The problem I have here is that it is hard to determine what is just a sub-menu and what is a sub-menu that also corresponds to a path. 

How about an option that takes a directory and uses it to construct a CMIS menu automatically. This would create entries for every directory/sub-directory from the root path (including save to folder entries like you suggested)? 

This would actually be quite handy. When I install CMIS I usually create a complex directory hierarchy and then set up the context menu one by one, which, in most cases is a long and arduous task.

> But I was expecting when the default Save Image As... is relocated to a sub-menu, it should also be removed from the main context menu.

Haha, yes, of course. I am working on an 'advanced options' tab. The ability to hide the default 'save image as' and 'save link as' will find their way there. Cheers.

**Comment by KevinSanders** [comment]
This is somewhat related, when creating a new folder path and you already have one selected on the left, it would be nice if it creates the new folder as a sub-folder of the currently selected one.

Also, I like that "Save here" as in the screengrab, could be a nice alternate way to save to the parent directory as I mentioned in a previous request.

**Comment by choobin** [comment]
> This is somewhat related, when creating a new folder path and you already have one selected on the left, it would be nice if it creates the new folder as a sub-folder of the currently selected one.

Good spot. From a usability perspective this feels much better. I have created a separate [issue](https://github.com/choobin/cmis/issues/16) for this so it does not get lost.

> Also, I like that "Save here" as in the screengrab, could be a nice alternate way to save to the parent directory as I mentioned in a previous request.

I have almost finished an option that will generate a menu from a directory hierarchy. This will create the 'Here' items for each directory by default. It hopefully should fulfill this requirement.

**Comment by choobin** [comment]
Save Image As menu items are now able to (optionally) select a directory to begin browsing from. Otherwise it will act as the default Firefox 'Save Image As' item. Furthermore, I have added an option to hide Firefox's "Save Image As" context menu item.

> Add an option to enable subfolders within an item.

Regarding this request. I have not explicitly added this, however, I have added a 'generate menu from directory tree' option. This will build a menu pretty much identical to the image nhantrn originally posted. 

Give them a whirl and see if they work out. I have just added a significant number of features and I would like to test them extensively over the weekend before sending off to AMO.
